cpp/src/arrow/compute/kernels/scalar_cast_nested.cc

```diff
@@ -150,6 +150,76 @@ void AddListCast(CastFunction* func) {
   DCHECK_OK(func->AddKernel(SrcType::type_id, std::move(kernel)));
 }
 
+struct CastStruct {
+  static Status Exec(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
+    const CastOptions& options = CastState::Get(ctx);
+    const auto in_field_count =
+        checked_cast<const StructType&>(*batch[0].type()).num_fields();
+    const auto out_field_count =
+        checked_cast<const StructType&>(*out->type()).num_fields();
+
+    if (in_field_count != out_field_count) {
+      return Status::TypeError("struct field sizes do not match: ", in_field_count,
+                               " and ", out_field_count);
+    }
```

Comment:
Sorry, looking at the output:

```
 [ RUN      ] ScanNode.MaterializationOfNestedVirtualColumn
/Users/runner/work/arrow/arrow/cpp/src/arrow/dataset/scanner_test.cc:1631: Failure
Value of: _st.IsNotImplemented()
  Actual: false
Expected: true
Expected 'fut.status()' to fail with NotImplemented, but got Type error: struct field sizes do not match: 1 and 2
/Users/runner/work/arrow/arrow/cpp/src/arrow/dataset/scanner_test.cc:1631: Failure
Value of: _st.ToString()
Expected: has substring "Unsupported cast from struct<e: int64> to struct"
  Actual: "Type error: struct field sizes do not match: 1 and 2"
[  FAILED  ] ScanNode.MaterializationOfNestedVirtualColumn (3 ms)
```

including the stringified types would probably be more helpful than the field count after all

Reply:
Also, that test needs adjusting^: we can just change it to expect TypeError instead of NotImplemented (we can implement the full path later)